### PR TITLE
Fiola merge

### PR DESCRIPTION
**Pull request by caichangjia**
# Description

Merge FIOLA into CaImAn

# Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
Add fiola class for real-time processing of fluorescence imaging data;
add demo_fiola.py file as demo;
add test_fiola.py file for tests;

### CODE REVIEW

File: caiman/fiola/gpu_mc_nnls.py [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+        self.template_zm, self.template_var = self.normalize_template(self.template[:,:,None,None], epsilon=self.epsilon)
+        self.shp = self.template.shape
+        self.shp_m_x, self.shp_m_y = self.shp[0]//2, self.shp[1]//2
+        self.target_freq = tf.signal.fft3d(tf.cast(self.template_zm[:,:,:,0], tf.complex128))
+        self.target_freq = tf.repeat(self.target_freq[None,:,:,0], repeats=[self.batch_size], axis=0)
+                
+        self.nc = self.template_0.shape[0]
+        self.nr = self.template_0.shape[1]
+        self.nd = 1
+        self.Nr = tf.signal.ifftshift(tf.range(-tf.experimental.numpy.fix(self.nr / 2.), tf.math.ceil(self.nr / 2.)))
```
[the remark this answers is not in the document]
Comment (pgunn) [reply caiman/fiola/gpu_mc_nnls.py:99]: If you can think of a way to do some abstraction to have this code be less repetitive (and operate less in-place), that'd be great.

File: caiman/fiola/gpu_mc_nnls.py [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+        # use gaussian interpolation to extract a fractional shift
+        shifts_int = self.argmax_2d(ncc) 
+        shifts_int_cast = tf.cast(shifts_int,tf.int32)
+        sh_x, sh_y = shifts_int_cast[:,0],shifts_int_cast[:,1]
+        sh_x_n = tf.cast(-(sh_x - ms_h), tf.float32)
+        sh_y_n = tf.cast(-(sh_y - ms_w), tf.float32)
+        ncc_log = tf.math.log(ncc)
+
+        n_batches = np.arange(self.batch_size)
+        idx = tf.transpose(tf.stack([n_batches, tf.squeeze(sh_x-1,axis=1), tf.squeeze(sh_y,axis=1)]))
```
Comment (pgunn) [inline caiman/fiola/gpu_mc_nnls.py:185]: Is there a more efficient, readable, and perhaps abstract way to write this block of code?

File: caiman/fiola/gpu_mc_nnls.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+        Returns:
+            x regressed values
+        
+        """
+        super().__init__(**kwargs)
+        self.th1 = theta_1.astype(np.float32)
+
+    def call(self, X):
+        """
+        pass as inputs the new Y, and the old X. see  https://angms.science/doc/NMF/nnls_pgd.pdf
```
Comment (pgunn) [inline caiman/fiola/gpu_mc_nnls.py:265]: document what it returns

File: caiman/fiola/signal_analysis_online.py [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+    idx = idx[np.logical_and(idx > (-window[0]), idx < (len(data) - window[-1]))]
+    PTD = data[(idx[:, np.newaxis] + window)]
+    PTA = np.median(PTD, 0)
+    
+    if template_window > 0:
+        # Template matching, second time thresholding
+        t_s = np.convolve(data, np.flipud(PTA), 'same')
+        data = t_s.copy()
+    else:
+        logging.info('skip template matching')
```
Comment (pgunn) [inline caiman/fiola/signal_analysis_online.py:391]: skipping template matching

File: caiman/fiola/signal_analysis_online.py [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+        std = np.sqrt(np.divide(np.sum(ff1**2), Ns)) 
+        
+    # Select best threshold based on estimated false positive rate
+    if adaptive_threshold:
+        pks2 = t_s[signal.find_peaks(t_s, height=None)[0]]
+        try:
+            thresh2, falsePosRate, detectionRate, low_spikes = adaptive_thresh(pks2, clip=0, pnorm=0.5, min_spikes=10)  # clip=0 means no clipping
+            thresh_factor = thresh2 / std
+            if thresh_factor < minimal_thresh:
+                logging.info(f'Adaptive threshold factor is lower than minimal theshold: {minimal_thresh}, choose thresh factor to be {minimal_thresh}')
```
Comment (pgunn) [inline caiman/fiola/signal_analysis_online.py:411]: Should this be an error or a warning? If it's just info people will usually not notice.

File: caiman/fiola/signal_analysis_online.py [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+                thresh_factor = minimal_thresh
+            thresh2 = thresh_factor * std
+        except:
+            logging.info('Adaptive threshold fails, automatically choose thresh factor to be 3')
```
Comment (pgunn) [inline caiman/fiola/signal_analysis_online.py:415]: Should this be an error?

File: caiman/fiola/signal_analysis_online.py [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+            thresh_factor = 3
+            thresh2 = thresh_factor * std
+    else:
+        raise ValueError('other methods are not implemented yet')
```
Comment (pgunn) [inline caiman/fiola/signal_analysis_online.py:419]: Prefer to say "Method {method} not supported" or something like that; user might've made a typo when requesting a method.

File: caiman/fiola/signal_analysis_online.py [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+
+    index = signal.find_peaks(data, height=thresh2)[0]
+    logging.info(f'final threshhold equals: {thresh2/std}')
```
Comment (pgunn) [inline caiman/fiola/signal_analysis_online.py:422]: "Final threshold: {thresh2/std}"

File: caiman/fiola/signal_analysis_online.py [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+        plt.title('template')
+        plt.show()
+
+        plt.figure()
+        plt.hist(pks2, 500)
+        plt.axvline(x=thresh2, c='r')
+        plt.title('distribution of spikes')
+        plt.tight_layout()
+        plt.show()  
+        plt.pause(5)
```
Comment (pgunn) [inline caiman/fiola/signal_analysis_online.py:451]: Probably fine, but I wonder if there's a better answer than an arbitrary pause of 5 seconds

File: caiman/fiola/signal_analysis_online.py [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+        
+        plt.figure();
+        plt.plot(PTA);
+        plt.title('template')
+        plt.show()
+
+        plt.figure()
+        plt.hist(pks2, 500)
+        plt.axvline(x=thresh2, c='r')
+        plt.title('distribution of spikes')
```
Comment (pgunn) [inline caiman/fiola/signal_analysis_online.py:448]: "Spike distribution"

File: caiman/tests/test_fiola.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        logging.warning('GPU is not detected')
+    fnames = download_demo('demo_voltage_imaging.hdf5', 'volpy')  # file path to movie file (will download if not present)
+    path_ROIs = download_demo('demo_voltage_imaging_ROIs.hdf5', 'volpy')  # file path to ROIs file (will download if not present)
+    mov = cm.load(fnames)
+    mask = cm.load(path_ROIs)
+    mov = mov[:2000]
+    
+    # setting params
+    # dataset dependent parameters
+    fnames = fnames                 # name of the movie, we don't put a name here as movie is already loaded above
```
Comment (pgunn) [inline caiman/tests/test_fiola.py:35]: This doesn't do anything

File: demos/general/demo_fiola.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+import glob
+import logging
+import matplotlib.pyplot as plt
+import numpy as np
+from numpy.linalg import norm
+import os
+os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
+from tensorflow.python.client import device_lib
+from time import time
+from threading import Thread
```
Comment (pgunn) [inline demos/general/demo_fiola.py:19]: Flip these lines to alphabetise

File: demos/general/demo_fiola.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+                    level=logging.INFO)
+
+logging.info(device_lib.list_local_devices())
+# if GPU is not detected, try to reinstall tensorflow with pip install tensorflow==2.4.1
+    
+#%%
+def main():
+    #%% load movie and masks
+    folder_idx = 0
+    movie_folder = ['/home/nel/NEL-LAB Dropbox/NEL/Papers/VolPy_online/data/voltage_data',
```
Comment (pgunn) [inline demos/general/demo_fiola.py:38]: These hardcoded paths are not going to be present

File: demos/general/demo_fiola.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    name = ['demo_voltage_imaging.hdf5', 'k53.tif'][folder_idx]
+    mode = ['voltage', 'calcium'][folder_idx]
+    
+    fnames = os.path.join(movie_folder, name)
+    mov = cm.load(os.path.join(movie_folder, name))
+    mask = cm.load(os.path.join(movie_folder, name.split('.')[0]+'_ROIs.hdf5'))
+    
+    display_images = False
+    if display_images:
+        plt.figure(); plt.imshow(mov.mean(0), vmax=np.percentile(mov.mean(0), 99.9)); plt.title('Mean img')
```
Comment (pgunn) [inline demos/general/demo_fiola.py:49]: separate lines

File: demos/general/demo_fiola.py [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        if mask is not None:
+            plt.figure(); plt.imshow(mask.mean(0)); plt.title('Masks')
```
Comment (pgunn) [inline demos/general/demo_fiola.py:51]: separate lines

File: demos/general/demo_fiola.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    params = fiolaparams(params_dict=options)
+    fio = FIOLA(params=params)
+    
+    #%% offline motion correction
+    fio.dims = mov.shape[1:]
+    template = mov.bin_median()
+    mc_mov, shifts, _ = fio.fit_gpu_motion_correction(mov, template, fio.params.mc_nnls['offline_batch_size'], mov.min())
+    
+    if display_images:
+        plt.figure(); plt.plot(shifts);plt.legend(['x shifts', 'y shifts']); plt.title('shifts'); plt.show()
```
Comment (pgunn) [inline demos/general/demo_fiola.py:67]: separate lines

File: demos/general/demo_fiola.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        fio.view_components(fio.corr)
+            
+    #%% save the result
+    save_name = f'{os.path.join(movie_folder, name.split(".")[0])}_fiola_result'
+    np.save(os.path.join(movie_folder, save_name), fio.estimates)
+    
+    #%%
+    log_files = glob.glob('*_LOG_*')
+    for log_file in log_files:
+        os.remove(log_file)
```
Comment (pgunn) [inline demos/general/demo_fiola.py:143]: Is removing log files something the user will usually want?

**Review by pgunn — CHANGES_REQUESTED** [review]
If you feel like it you could also add typing to the function declarations, although if you don't want to do that, it's fine. You don't need to address all my comments.

My reasoning behind preferring log phrasing like "Doing this" rather than "Do this" is that using the imperative voice *only* when the user is being asked to do something helps distinguish "User must do this" from "program is currently doing this".

File: caiman/fiola/gpu_mc_nnls.py [hunk 18]
```diff
@@ -0,0 +1,605 @@
...
+Created on Mon Aug 30 11:01:52 2021
+This file includes class for gpu motion correction, source extraction(NNLS) and
+pipeline which process motion correction, source extraction, spike detection frame
+by-frame.
+@author: @caichangjia, @cynthia, @agiovann
+"""
+import logging
+import numpy as np
+import os
+os.environ["CUDA_VISIBLE_DEVICES"] = "0";
```
[the remark this answers is not in the document]
Comment (caichangjia) [reply caiman/fiola/gpu_mc_nnls.py:13]: it sets particular gpu for running the code, but seems not necessary in our case

**Comment by pgunn** [comment]
Another "would be nice" would be to use type annotations for arguments and return values, although I can add this in afterwards if you prefer.

File: caiman/fiola/fiola.py [hunk 19]
```diff
@@ -0,0 +1,657 @@
...
+@author: @agiovann, @caichangjia, @cynthia
+"""
+import logging
+import matplotlib.pyplot as plt
+from matplotlib.widgets import Slider
+import numpy as np
+from numpy.linalg import norm
+from scipy.optimize import nnls  
+import tensorflow as tf
+gpus = tf.config.experimental.list_physical_devices('GPU')
```
[the remark this answers is not in the document]
Comment (caichangjia) [reply caiman/fiola/fiola.py:17]: It is the default way to limit GPU memory in tensorflow2 and it is necessary for our code. I don't find an alternative unfortunately.

**Review by pgunn — APPROVED** [review]

**Pull request merged**